src/compiler/checker.ts

```diff
@@ -6785,30 +6830,175 @@ module ts {
                     // Check spread elements against rest type (from arity check we know spread argument corresponds to a rest parameter)
                     let paramType = getTypeAtPosition(signature, i);
-                    // A tagged template expression provides a special first argument, and string literals get string literal types
+
+                    // Decorators provide special arguments, a tagged template expression provides 
+                    // a special first argument, and string literals get string literal types
                     // unless we're reporting errors
-                    let argType = i === 0 && node.kind === SyntaxKind.TaggedTemplateExpression
-                        ? globalTemplateStringsArrayType
-                        : arg.kind === SyntaxKind.StringLiteral && !reportErrors
-                            ? getStringLiteralType(<StringLiteral>arg)
-                            : checkExpressionWithContextualType(arg, paramType, excludeArgument && excludeArgument[i] ? identityMapper : undefined);
+                    let argType: Type;
+                    if (arg.parent.kind === SyntaxKind.Decorator) {
+                        if (i === 0) {
+                            argType = getDecoratorTargetArgumentType(arg.parent.parent);
+                        }
+                        else if (i === 1) {
+                            argType = getDecoratorPropertyKeyArgumentType(arg.parent.parent);
+                        }
+                        else if (i === 2) {
+                            argType = getDecoratorPropertyIndexOrDescriptorArgumentType(arg.parent.parent);
+                        }
+                    }
+                    else if (i === 0 && arg.parent.kind === SyntaxKind.TaggedTemplateExpression) {
+                        argType = globalTemplateStringsArrayType;
+                    }
+
+                    if (argType === undefined) {
+                        if (arg.kind === SyntaxKind.StringLiteral && !reportErrors) {
+                            argType = getStringLiteralType(<StringLiteral>arg);
+                        }
+                        else {
+                            argType = checkExpressionWithContextualType(arg, paramType, excludeArgument && excludeArgument[i] ? identityMapper : undefined);
+                        }
+                    }
 
                     // Use argument expression as error location when reporting errors
-                    if (!checkTypeRelatedTo(argType, paramType, relation, reportErrors ? arg : undefined,
-                        Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1)) {
+                    let errorNode = reportErrors ? arg : undefined;
+                    let headMessage = Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1;
+                    if (!checkTypeRelatedTo(argType, paramType, relation, errorNode, headMessage, containingMessageChain)) {

```

Comment:
Sorry, the part that I originally found "gross" was that this call was ever in the `if` in the first place, not your changes; it's more readable now, but if you feel that this is reasonable, we can roll with this.